docs/functionaldatatypes/validation.md

```diff
@@ -0,0 +1,286 @@
+
+However, it is not what we want here.
+
+We could imagine that the `name` and `age` were transmitted by an internal business customer. They would probably not be happy if we told them the name was invalid, they corrected that and sent it again, and then we failed again because the age was invalid.
+
+"Why didn't you tell me about all the problems up front?" we can see them saying. `Validation` does just that.
+
+We can fix this example using `Validation` like so:
+
+```scala mdoc:reset
+import zio.prelude.Validation
+
+case class Person(name: String, age: Int)
+
+def validateName(name: String): Validation[String, String] =
+  if (name.isEmpty) Validation.fail("Name was empty")
+  else Validation.succeed(name)
+
+def validateAge(age: Int): Validation[String, Int] =
+  if (age <= 0) Validation.fail(s"Age $age was less than zero")
+  else Validation.succeed(age)
+
+def validatePerson(name: String, age: Int): Validation[String, Person] =
+  Validation.validateWith(validateName(name), validateAge(age))(Person)
+```
+
+Here we are just using the basic operators on `Validation` of `succeed` and `fail` which construct a `Validation` that succeeds or fails with the specified value. We are then using the `validateWith` operator, which combines the results of multiple `Validation` values, accumulating all errors that occur.
+
+Note that we are using the `apply` method of `Person` to combine the validated name and age into a `Person`. This is often a convenient pattern when we are modeling our data using case classes.
+
+Now if we validate a person where both the `name` and the `age` are invalid we will get a failure containing both errors, giving us exactly what we need for our internal customer so they can efficiently correct the problem and send us valid data.
+
+## Constructing Validation Values
+
+Now that we understand what `Validation` is, let's look at how to construct `Validation` values.
+
+### From Existing Values
+
+The simplest ways to construct `Validation` values are the `succeed` and `fail` operators we saw above.
+
+```scala mdoc:nest
+object Validation {
+  def fail[E](error: E): Validation[E, Nothing] =
+    ???
+  def succeed[A](value: A): Validation[Nothing, A] =
+    ???
+}
+```
+
+These just construct validation successes or failures with the specified value. We can use these as the building blocks for creating more complex validation logic as we saw above.
+
+### From Code That May Throw Exceptions
+
+Another useful constructor of `Validation` values is the `apply` method on `Validation`.
+
+```scala mdoc:nest
+object Validation {
+  def apply[A](a: => A): Validation[Throwable, A] =
+    ???
+}
+```
+
+The `apply` operator takes a by name argument and evaluates that argument, converting any non-fatal thrown exceptions into `Validation` failures. This is useful when working with code that might throw exceptions to safely import that code into a `Validation` value.
```

Comment:
In practice, I've found this very useful recently.

Reply:
Good feedback.